core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,243 @@
+pub trait ReadStorageFactory: Clone + Debug + Send + Sync {
+    type ReadStorageImpl<'a>: ReadStorage
+    where
+        Self: 'a;
+
+    async fn access_storage<'a>(
+        &'a self,
+        rt_handle: Handle,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> Option<Self::ReadStorageImpl<'a>>;
+}
+
+/// A [`ReadStorage`] implementation that uses either [`PostgresStorage`] or [`RocksdbStorage`]
+/// underneath.
+#[derive(Debug)]
+pub enum PgOrRocksdbStorage<'a> {
+    Postgres(PostgresStorage<'a>),
+    RocksdbStorage(RocksdbStorage),
+}
+
+impl<'a> ReadStorage for PgOrRocksdbStorage<'a> {
+    fn read_value(&mut self, key: &zksync_types::StorageKey) -> zksync_types::StorageValue {
+        match self {
+            PgOrRocksdbStorage::Postgres(postgres) => postgres.read_value(key),
+            PgOrRocksdbStorage::RocksdbStorage(rocksdb) => rocksdb.read_value(key),
+        }
+    }
+
+    fn is_write_initial(&mut self, key: &zksync_types::StorageKey) -> bool {
+        match self {
+            PgOrRocksdbStorage::Postgres(postgres) => postgres.is_write_initial(key),
+            PgOrRocksdbStorage::RocksdbStorage(rocksdb) => rocksdb.is_write_initial(key),
+        }
+    }
+
+    fn load_factory_dep(&mut self, hash: zksync_types::H256) -> Option<Vec<u8>> {
+        match self {
+            PgOrRocksdbStorage::Postgres(postgres) => postgres.load_factory_dep(hash),
+            PgOrRocksdbStorage::RocksdbStorage(rocksdb) => rocksdb.load_factory_dep(hash),
+        }
+    }
+
+    fn get_enumeration_index(&mut self, key: &zksync_types::StorageKey) -> Option<u64> {
+        match self {
+            PgOrRocksdbStorage::Postgres(postgres) => postgres.get_enumeration_index(key),
+            PgOrRocksdbStorage::RocksdbStorage(rocksdb) => rocksdb.get_enumeration_index(key),
+        }
+    }
+}
+
+impl<'a> From<PostgresStorage<'a>> for PgOrRocksdbStorage<'a> {
+    fn from(value: PostgresStorage<'a>) -> Self {
+        PgOrRocksdbStorage::Postgres(value)
+    }
+}
+
+impl<'a> From<RocksdbStorage> for PgOrRocksdbStorage<'a> {
+    fn from(value: RocksdbStorage) -> Self {
+        PgOrRocksdbStorage::RocksdbStorage(value)
+    }
+}
+
+/// A [`ReadStorageFactory`] implementation that can produce short-lived [`ReadStorage`] handles
+/// backed by either Postgres or RocksDB (if it's caught up). Always initialized as a `Postgres`
+/// variant and is then mutated into `Rocksdb` once RocksDB cache is caught up. After which it
+/// can never revert back to `Postgres` as we assume RocksDB cannot fall behind under normal state
+/// keeper operation.
+#[derive(Debug, Clone)]
+pub enum AsyncRocksdbCache {
+    Postgres(ConnectionPool),
+    Rocksdb(RocksDB<StateKeeperColumnFamily>, ConnectionPool),
+}
+
+impl AsyncRocksdbCache {
+    /// Returns a [`ReadStorage`] implementation backed by Postgres
+    async fn access_storage_pg(
+        rt_handle: Handle,
+        pool: &ConnectionPool,
+    ) -> anyhow::Result<PgOrRocksdbStorage> {
```

Comment:
Please use `<'_>` for `PgOrRocksdbStorage` to cleanly specify that it has a lifetime.